openff/toolkit/topology/topology.py

```diff
@@ -1180,23 +1171,105 @@ def chemical_environment_matches(
             if len(mol_matches) == 0:
                 continue
 
-            # Loop over matches
-            for match in mol_matches:
+            for mol_instance_idx, atom_map in group:
+                mol_instance = self.molecule(mol_instance_idx)
+                # Loop over matches
+                for match in mol_matches:
 
-                # Collect indices of matching TopologyAtoms.
-                topology_atom_indices = []
-                for molecule_atom_index in match:
-                    reference_atom = molecule.atoms[molecule_atom_index]
-                    # topology_atom = TopologyAtom(reference_atom, topology_molecule)
-                    topology_atom_indices.append(self.atom_index(reference_atom))
+                    # Collect indices of matching TopologyAtoms.
+                    topology_atom_indices = []
+                    for molecule_atom_index in match:
+                        atom = mol_instance.atom(atom_map[molecule_atom_index])
+                        topology_atom_indices.append(self.atom_index(atom))
 
-                environment_match = Topology._ChemicalEnvironmentMatch(
-                    tuple(match), molecule, tuple(topology_atom_indices)
-                )
+                    environment_match = Topology._ChemicalEnvironmentMatch(
+                        tuple(match), unique_mol, tuple(topology_atom_indices)
+                    )
 
-                matches.append(environment_match)
+                    matches.append(environment_match)
         return matches
 
+    @property
+    def identical_molecule_groups(self):
+        """
+        Returns groups of chemically identical molecules, identified by index and atom map.
+
+        Returns
+        -------
+        identical_molecule_groups : {int:[[int: {int: int}]]}
+            A dict of the form {unique_mol_idx : [[topology_mol_idx, atom_map],...].
+            A dict where each key is the topology molecule index of a unique chemical species, and each value is a list
+            describing all of the instances of that chemical species in the topology. Each instance is a
+            two-membered list where the first element is the topology molecule index, and the second element
+            is a dict describing the atom map from the unique molecule to the instance of it in the topology.
+
```

Comment:
This docstring could use some work - not easy to encode the complex data type returned here.